pennylane/math/matrix_manipulation.py

```diff
@@ -104,44 +106,99 @@ def expand_matrix(base_matrix, wires, wire_order=None, sparse_format="csr"):
-    n = len(wires)
+    wire_order = list(wire_order)
+
     shape = qml.math.shape(base_matrix)
     batch_dim = shape[0] if len(shape) == 3 else None
 
-    # operator's wire positions relative to wire ordering
-    op_wire_pos = wire_order.indices(wires)
-
-    identity = qml.math.reshape(
-        qml.math.eye(2 ** len(wire_order), like=interface), [2] * (len(wire_order) * 2)
-    )
-    # The first axis entries are range(n, 2n) for batch_dim=None and range(n+1, 2n+1) else
-    axes = (list(range(-n, 0)), op_wire_pos)
-
-    # reshape op.matrix()
-    op_matrix_interface = qml.math.convert_like(base_matrix, identity)
-    shape = [batch_dim] + [2] * (n * 2) if batch_dim else [2] * (n * 2)
-    mat_op_reshaped = qml.math.reshape(op_matrix_interface, shape)
-    mat_tensordot = qml.math.tensordot(
-        mat_op_reshaped, qml.math.cast_like(identity, mat_op_reshaped), axes
-    )
-
-    unused_idxs = [idx for idx in range(len(wire_order)) if idx not in op_wire_pos]
-    # permute matrix axes to match wire ordering
-    perm = op_wire_pos + unused_idxs
-    sources = wire_order.indices(wire_order)
+    # get a subset of `wire_order` values that contain all wire labels inside `wires` argument
+    # e.g. wire_order = [0, 1, 2, 3, 4]; wires = [3, 0, 2]
+    # --> subset_wire_order = [0, 1, 2, 3]; expanded_wires = [3, 0, 2, 1]
+    wires = list(wires)
+    wire_indices = [wire_order.index(wire) for wire in wires]
+    subset_wire_order = wire_order[min(wire_indices) : max(wire_indices) + 1]
+    wire_difference = list(set(subset_wire_order) - set(wires))
+    expanded_wires = wires + wire_difference
+
+    # expand matrix if needed
+    if wire_difference:
+        if batch_dim is not None:
+            batch_matrices = [
+                qml.math.kron(batch, eye_interface(len(wire_difference)), like=interface)
+                for batch in base_matrix
+            ]
+            base_matrix = qml.math.stack(batch_matrices, like=interface)
+        else:
+            base_matrix = qml.math.kron(
+                base_matrix, eye_interface(len(wire_difference)), like=interface
+            )
+
+    num_wires = len(subset_wire_order)
+
+    # reshape matrix to match wire values e.g. mat[0, 0, 0, 0] = <0000|mat|0000>
+    # with this reshape we can easily swap wires
+    shape = [batch_dim] + [2] * (num_wires * 2) if batch_dim else [2] * (num_wires * 2)
+    mat = qml.math.reshape(base_matrix, shape)
+
+    # compute the permutations needed to match wire order
+    perm = [expanded_wires.index(wire) for wire in subset_wire_order]
+    perm += [p + num_wires for p in perm]
     if batch_dim:
-        perm = [p + 1 for p in perm]
-        sources = [s + 1 for s in sources]
+        perm = [0] + [p + 1 for p in perm]
 
-    mat = qml.math.moveaxis(mat_tensordot, sources, perm)
-    shape = [batch_dim] + [2 ** len(wire_order)] * 2 if batch_dim else [2 ** len(wire_order)] * 2
+    if perm != list(range(num_wires * 2)):
```

Comment:
If we don't need to perform re-order the dimensions, we don't need to reshape in the first place.  How about moving the reshaping inside?

Another point, does this check work when there's a batched dimension?

Reply:
You are right! Moved code inside the if statement.

Reply:
Changed check to `expanded_wires != subset_wire_order` to make sure it works for batched dimensions.